### PR TITLE
Use `systemd` in test containers

### PR DESCRIPTION
**Pull request by yubiuser**
**What does this PR aim to accomplish?:**

PR https://github.com/pi-hole/pi-hole/pull/4924 added a native `systemd` service file. All our currently supported OS run systemd. However, during the automatic test suite we don't run `systemd` so far and mock `systemctl` call.

https://github.com/pi-hole/pi-hole/blob/0034538794a3754ab532a53480da39d3fd13d7ed/test/test_any_automated_install.py#L172-L181

Some images already require installation of an initsystem (e.g. centoOS and Fedora). This PR adds `systemd`as initsystem to the images so that the containers start with `systemd` as PID 1. This mimics the real-world OS situation better.

**How does this PR accomplish the above?:**

1. Adds the required `systemd` packages to the images.
2. ~~Starts the test containers with `--privileged` instead of `--cap-add=ALL` (required to start `systemd` within the container)~~ Uses `podman` instead of `docker` as container manager
3. Removes the mocked `systemctl` from the test suite.
4. Allows `pihole restartdns` to use `systemctl` instead of `service` if available. See this PR https://github.com/pi-hole/pi-hole/pull/5045 for more details

---
**By submitting this pull request, I confirm the following:**

1. I have read and understood the [contributors guide](https://docs.pi-hole.net/guides/github/contributing/), as well as this entire template. I understand which branch to base my commits and Pull Requests against.
2. I have commented my proposed changes within the code and I have tested my changes.
5. I am willing to help maintain this change if there are issues with it later.
6. It is compatible with the [EUPL 1.2 license](https://opensource.org/licenses/EUPL-1.1)
7. I have squashed any insignificant commits. ([`git rebase`](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))
8. I have checked that another pull request for this purpose does not exist.
9. I have considered, and confirmed that this submission will be valuable to others.
10. I accept that this submission may not be used, and the pull request closed at the will of the maintainer.
11. I give this submission freely, and claim no ownership to its content.

---
- [x] I have read the above and my PR is ready for review. *Check this box to confirm*

### CODE REVIEW

**Comment by yubiuser** [comment]
I removed the mocked `systemctl` command in https://github.com/pi-hole/pi-hole/pull/5044/commits/8e6bb152831c291d5d0afb47f3a565bab3a47e7b. However, we could leave it in there and bypass it by checking if `systemd` is PID 1. This would keep the code in case we want to add a OS which is not using `systemd` as init system in the future.

**Comment by yubiuser** [comment]
I'll make a separate PR from https://github.com/pi-hole/pi-hole/pull/5044/commits/097e758cb602e897e2e497099e75c187c26d3d7b. Just leaving it in here at the moment as it is a requirement for some OS to pass the test. I'll rebase once the other PR is merged.

**Comment by yubiuser** [comment]
Converting to draft until I've figured out why CentOS 8 needs 2:45h to pass the test on GH. Runs fine locally within 5 mins.

**Comment by PromoFaux** [comment]
According to the logs this is still taking a long time for Centos8 - have you got to the bottom of why?

**Comment by DL6ER** [comment]
> Allows pihole restartdns to use systemctl instead of service if available

I don't think this is necessary as `service` is a wrapper. Do we really need this (complexity)?

**Comment by yubiuser** [comment]
I think yes, that's why I created the PR.

All distros use `systemd` by default, we have a native `systemd` unit now. Some container images already need an additional `initsystem` to pass the test (fedora, centos) - there is no `service` installed. Instead of installing `sysVinit` we should test with what is actually used by the distros.
Background of all of this: We use the `service` file (and now `prestart.sh`) to set file permissions and create files/folders. We check for them. So we need to start the scripts/service to set the permissions.

**Comment by yubiuser** [comment]
I just pushed the identical commits to a new branch and created PR https://github.com/pi-hole/pi-hole/pull/5057 and it finished in 4 minutes for CentOS8 
https://github.com/pi-hole/pi-hole/actions/runs/3670895068/jobs/6205732893

**Comment by PromoFaux** [comment]
Maybe there was just something wrong a week ago and it just needed a kick today. I think we can close the other PR.

>Do we really need this (complexity)?

One thing I do like here is that it removes some of the clusterfuck code in the tests that mocks systemctl, and anything we can do to make the tests easier to follow and maintain is a good thing in my book.

File: test/_centos_8.Dockerfile [hunk 1]
```diff
@@ -1,5 +1,13 @@
...
-RUN yum install -y git initscripts
+RUN yum install -y git systemd
+
+RUN rm -f /lib/systemd/system/multi-user.target.wants/* \
+    /etc/systemd/system/*.wants/* \
+    /lib/systemd/system/local-fs.target.wants/* \
+    /lib/systemd/system/sockets.target.wants/*udev* \
+    /lib/systemd/system/sockets.target.wants/*initctl* \
+    /lib/systemd/system/sysinit.target.wants/systemd-tmpfiles-setup* \
+    /lib/systemd/system/systemd-update-utmp*
```
Comment (MichaIng) [inline test/_centos_8.Dockerfile:10]: Does `systemd` not detect the container environment and skips conflicting units automatically? Would be closer to real-world testing if systemd is not prevented from starting anything (including its core components) but FTL.

Comment (yubiuser) [reply test/_centos_8.Dockerfile]: > Does systemd not detect the container environment and skips conflicting units automatically? 

I don't know. 
I setup `systemd` in the containers in a way I read somewhere. It made sense to remove unwanted units which are not needed for our own service. However, I see your point: we are not using `docker` here as a way to start a single service but as a proxy for real full-blown systems. I removed the cleaning.

Comment (yubiuser) [reply test/_centos_8.Dockerfile]: Mhh...

> Error: URL https://github.com/pi-hole/ftl/releases/latest/download/pihole-FTL-linux-x86_64 not found

Comment (yubiuser) [reply test/_centos_8.Dockerfile]: Locally, tests pass on `debian_11`

**Comment by yubiuser** [comment]
After random fails occurred on github (and later also locally) I decided to take a different approach: My suspicion was that I needed to start the container with `--privileged` and didn't know what the container might does to the host and if it influenced it's networking (usually tests failed while downloading `FTL`binary). There was no way I could make `systemd` work inside a docker container without `--privileged`.
Background: `systemd` and `docker` don't play well together - each is blaming the other one. Esp. since a `systemd` 248 (affecting `cgroup` handling) things got worse: https://github.com/systemd/systemd/issues/19245. Systemd published a container interface documentation (https://systemd.io/CONTAINER_INTERFACE/), however, this is not well supported by `docker`.

However, other container manager play better with `systemd` - e.g. `podman`. This is developed by RedHat and meant as a `docker` in-place replacement. Unlike `docker` it not a daemon based system and runs rootless by default. Bonus: it supports `systemd` inside the container out-of-the-box. This way, I didn't need the `--privileged` but could leave capabilities untouched.

Switching from `docker` to `podman` was easier than I first thought: it's installed by default on ubuntu runners (https://github.com/actions/runner-images/blob/main/images/linux/Ubuntu2204-Readme.md) and `testinfra` has a `podman` backend.

After switching, test have not failed so far.

**Comment by MichaIng** [comment]
Another idea would be [systemd-nspawn](https://manpages.debian.org/systemd-nspawn), which can launch the init system of any OS directory tree (e.g. just generated with `debootstrap`) or loop-mounted OS image. Simple compared to Docker, more an advanced chroot with filesystem and process tree virtualization, optional mounts from host, optional internal networking etc.

However, too much change away from Docker for now, just to keep in mind that there are other/simpler container engines with native systemd support, when needed.

**Comment by PromoFaux** [comment]
> Switching from docker to podman was easier than I first thought

My only concern here is it's another thing I need to install in order for the tests to run. 

That said, even after installing `podman`, I am faced with all tests failing:

```
adam@adam-pc:~/pi-hole$ podman --version
podman version 3.4.4
```

Should note that this is a WSL2 instance of Ubuntu 22.04

```
================================================================================= short test summary info ==================================================================================
ERROR test_any_automated_install.py::test_package_manager_has_pihole_deps - subprocess.CalledProcessError: Command '['podman', 'run', '-t', '-d', '--cap-add=ALL', 'pytest_pihole:test_container']' returned non-zero exit status 127.
ERROR test_any_automated_install.py::test_os_check_passes - subprocess.CalledProcessError: Command '['podman', 'run', '-t', '-d', '--cap-add=ALL', 'pytest_pihole:test_container']' returned non-zero exit status 127.
ERROR test_any_automated_install.py::test_FTL_detect_armv4t_no_errors - subprocess.CalledProcessError: Command '['podman', 'run', '-t', '-d', '--cap-add=ALL', 'pytest_pihole:test_container']' returned non-zero exit status 127.
ERROR test_any_automated_install.py::test_installPihole_fresh_install_readableBlockpage[True] - subprocess.CalledProcessError: Command '['podman', 'run', '-t', '-d', '--cap-add=ALL', 'pytest_pihole:test_container']' returned non-zero exit status 127.
ERROR test_any_automated_install.py::test_IPv6_only_ULA - subprocess.CalledProcessError: Command '['podman', 'run', '-t', '-d', '--cap-add=ALL', 'pytest_pihole:test_container']' returned non-zero exit status 127.
ERROR test_any_automated_install.py::test_installPiholeWeb_fresh_install_no_errors - subprocess.CalledProcessError: Command '['podman', 'run', '-t', '-d', '--cap-add=ALL', 'pytest_pihole:test_container']' returned non-zero exit status 127.
ERROR test_any_automated_install.py::test_validate_ip - subprocess.CalledProcessError: Command '['podman', 'run', '-t', '-d', '--cap-add=ALL', 'pytest_pihole:test_container']' returned non-zero exit status 127.
ERROR test_any_automated_install.py::test_supported_package_manager - subprocess.CalledProcessError: Command '['podman', 'run', '-t', '-d', '--cap-add=ALL', 'pytest_pihole:test_container']' returned non-zero exit status 127.
ERROR test_any_automated_install.py::test_FTL_binary_installed_and_responsive_no_errors - subprocess.CalledProcessError: Command '['podman', 'run', '-t', '-d', '--cap-add=ALL', 'pytest_pihole:test_container']' returned non-zero exit status 127.
ERROR test_any_automated_install.py::test_setupVars_saved_to_file - subprocess.CalledProcessError: Command '['podman', 'run', '-t', '-d', '--cap-add=ALL', 'pytest_pihole:test_container']' returned non-zero exit status 127.
ERROR test_any_utils.py::test_key_val_replacement_works - subprocess.CalledProcessError: Command '['podman', 'run', '-t', '-d', '--cap-add=ALL', 'pytest_pihole:test_container']' returned non-zero exit status 127.
ERROR test_any_automated_install.py::test_IPv6_GUA_ULA_test - subprocess.CalledProcessError: Command '['podman', 'run', '-t', '-d', '--cap-add=ALL', 'pytest_pihole:test_container']' returned non-zero exit status 127.
ERROR test_any_automated_install.py::test_FTL_detect_armv6l_no_errors - subprocess.CalledProcessError: Command '['podman', 'run', '-t', '-d', '--cap-add=ALL', 'pytest_pihole:test_container']' returned non-zero exit status 127.
ERROR test_any_automated_install.py::test_FTL_detect_unknown_no_errors - subprocess.CalledProcessError: Command '['podman', 'run', '-t', '-d', '--cap-add=ALL', 'pytest_pihole:test_container']' returned non-zero exit status 127.
ERROR test_any_automated_install.py::test_FTL_detect_armv8a_no_errors - subprocess.CalledProcessError: Command '['podman', 'run', '-t', '-d', '--cap-add=ALL', 'pytest_pihole:test_container']' returned non-zero exit status 127.
ERROR test_any_automated_install.py::test_update_package_cache_failure_no_errors - subprocess.CalledProcessError: Command '['podman', 'run', '-t', '-d', '--cap-add=ALL', 'pytest_pihole:test_container']' returned non-zero exit status 127.
ERROR test_any_automated_install.py::test_package_manager_has_web_deps - subprocess.CalledProcessError: Command '['podman', 'run', '-t', '-d', '--cap-add=ALL', 'pytest_pihole:test_container']' returned non-zero exit status 127.
ERROR test_any_automated_install.py::test_package_manager_has_installer_deps - subprocess.CalledProcessError: Command '['podman', 'run', '-t', '-d', '--cap-add=ALL', 'pytest_pihole:test_container']' returned non-zero exit status 127.
ERROR test_any_automated_install.py::test_FTL_detect_armv5te_no_errors - subprocess.CalledProcessError: Command '['podman', 'run', '-t', '-d', '--cap-add=ALL', 'pytest_pihole:test_container']' returned non-zero exit status 127.
ERROR test_any_automated_install.py::test_update_package_cache_success_no_errors - subprocess.CalledProcessError: Command '['podman', 'run', '-t', '-d', '--cap-add=ALL', 'pytest_pihole:test_container']' returned non-zero exit status 127.
ERROR test_any_automated_install.py::test_installPihole_fresh_install_readableFiles - subprocess.CalledProcessError: Command '['podman', 'run', '-t', '-d', '--cap-add=ALL', 'pytest_pihole:test_container']' returned non-zero exit status 127.
ERROR test_any_automated_install.py::test_IPv6_only_GUA - subprocess.CalledProcessError: Command '['podman', 'run', '-t', '-d', '--cap-add=ALL', 'pytest_pihole:test_container']' returned non-zero exit status 127.
ERROR test_any_automated_install.py::test_setupVars_are_sourced_to_global_scope - subprocess.CalledProcessError: Command '['podman', 'run', '-t', '-d', '--cap-add=ALL', 'pytest_pihole:test_container']' returned non-zero exit status 127.
ERROR test_any_automated_install.py::test_selinux_not_detected - subprocess.CalledProcessError: Command '['podman', 'run', '-t', '-d', '--cap-add=ALL', 'pytest_pihole:test_container']' returned non-zero exit status 127.
ERROR test_any_automated_install.py::test_os_check_fails - subprocess.CalledProcessError: Command '['podman', 'run', '-t', '-d', '--cap-add=ALL', 'pytest_pihole:test_container']' returned non-zero exit status 127.
ERROR test_any_automated_install.py::test_FTL_download_aarch64_no_errors - subprocess.CalledProcessError: Command '['podman', 'run', '-t', '-d', '--cap-add=ALL', 'pytest_pihole:test_container']' returned non-zero exit status 127.
ERROR test_any_automated_install.py::test_IPv6_ULA_GUA_test - subprocess.CalledProcessError: Command '['podman', 'run', '-t', '-d', '--cap-add=ALL', 'pytest_pihole:test_container']' returned non-zero exit status 127.
ERROR test_any_automated_install.py::test_FTL_detect_armv7l_no_errors - subprocess.CalledProcessError: Command '['podman', 'run', '-t', '-d', '--cap-add=ALL', 'pytest_pihole:test_container']' returned non-zero exit status 127.
ERROR test_any_automated_install.py::test_FTL_detect_x86_64_no_errors - subprocess.CalledProcessError: Command '['podman', 'run', '-t', '-d', '--cap-add=ALL', 'pytest_pihole:test_container']' returned non-zero exit status 127.
ERROR test_any_utils.py::test_key_addition_works - subprocess.CalledProcessError: Command '['podman', 'run', '-t', '-d', '--cap-add=ALL', 'pytest_pihole:test_container']' returned non-zero exit status 127.
ERROR test_any_utils.py::test_key_removal_works - subprocess.CalledProcessError: Command '['podman', 'run', '-t', '-d', '--cap-add=ALL', 'pytest_pihole:test_container']' returned non-zero exit status 127.
ERROR test_any_utils.py::test_getFTLAPIPort_default - subprocess.CalledProcessError: Command '['podman', 'run', '-t', '-d', '--cap-add=ALL', 'pytest_pihole:test_container']' returned non-zero exit status 127.
ERROR test_any_utils.py::test_getFTLAPIPort_custom - subprocess.CalledProcessError: Command '['podman', 'run', '-t', '-d', '--cap-add=ALL', 'pytest_pihole:test_container']' returned non-zero exit status 127.
ERROR test_any_utils.py::test_getFTLPID_default - subprocess.CalledProcessError: Command '['podman', 'run', '-t', '-d', '--cap-add=ALL', 'pytest_pihole:test_container']' returned non-zero exit status 127.
ERROR test_any_automated_install.py::test_FTL_detect_aarch64_no_errors - subprocess.CalledProcessError: Command '['podman', 'run', '-t', '-d', '--cap-add=ALL', 'pytest_pihole:test_container']' returned non-zero exit status 127.
ERROR test_any_utils.py::test_getFTLAPIPort_malicious - subprocess.CalledProcessError: Command '['podman', 'run', '-t', '-d', '--cap-add=ALL', 'pytest_pihole:test_container']' returned non-zero exit status 127.
ERROR test_any_utils.py::test_getFTLPIDFile_default - subprocess.CalledProcessError: Command '['podman', 'run', '-t', '-d', '--cap-add=ALL', 'pytest_pihole:test_container']' returned non-zero exit status 127.
ERROR test_any_automated_install.py::test_IPv6_only_link_local - subprocess.CalledProcessError: Command '['podman', 'run', '-t', '-d', '--cap-add=ALL', 'pytest_pihole:test_container']' returned non-zero exit status 127.
ERROR test_any_utils.py::test_getFTLPIDFile_and_getFTLPID_custom - subprocess.CalledProcessError: Command '['podman', 'run', '-t', '-d', '--cap-add=ALL', 'pytest_pihole:test_container']' returned non-zero exit status 127.
==================================================================================== 39 errors in 5.28s ====================================================================================
py3: exit 1 (5.48 seconds) /home/adam/pi-hole/test> pytest -vv -n auto ./test_any_automated_install.py ./test_any_utils.py pid=3675
.pkg: _exit /home/adam/pi-hole/test> python /home/adam/.local/lib/python3.10/site-packages/pyproject_api/_backend.py True setuptools.build_meta __legacy__
  py3: FAIL code 1 (19.04=setup[1.40]+cmd[12.16,5.48] seconds)
  evaluation failed :( (19.07 seconds)
```

I have not yet installed `podman` natively in windows to try there... but I would rather not have to at this stage.

**Pull request closed without merging**